CMakeLists.txt

```diff
@@ -287,6 +287,13 @@ zephyr_ld_options(
   ${LINKERFLAGPREFIX},--build-id=none
   )
 
+# Funny thing is if you set this to =error, some architectures will
+# fail the compiler flag check and this will be removed from build.
+# So keep it at =warn.
```

Comment:
Would

zephyr_link_libraries(
  ${LINKERFLAGPREFIX},--orphan-handling=error
  )

have the behaviour we want?

Reply:
The linker will fail if =error is used when there are orphan sections. However, the ARC and Xtensa, cmake somehow silently drops the flag. There is nothing to indicate problems in the CMakeError.log. Deliberately setting the wrong parameter (e.g. =badarg) would result in error output about invalid argument in CMakeError.log. I haven't been able to figure out why cmake would drop the flag. So we can only use =warn for now to enable warning for all architectures. Sanitycheck will still errors out if there are any orphan sections.

Reply:
I was able to reproduce your observation about the flag being dropped with zephyr_ld_options, but when using zephyr_link_libraries I got the desired behaviour. Error on orphan sections.

Note, using zephyr_link_libraries should be used in place of zephyr_ld_options.

zephyr_ld_options will by-design drop the flag if cmake is unable to compile a simple C file with it (presumably there are some orphan sections in the compiler flag test).

zephyr_link_libraries just adds the flag without any toolchain-capability test.

https://gist.github.com/SebastianBoe/d480b62d0f998d0f108ad75fba8ed1fc

zephyr-sdk-0.9.3
qemu_xtensa

Reply:
I thought about it but I am hesitant on forcing this flag without compiler check. Putting this in zephyr_link_libraries assumes everyone will be using binutils to link the final binaries. As we support custom toolchain, that compiler/linker combo may not support this flag.